### PR TITLE
Sketcher : ArcLength Constraint

### LINKED ISSUE #9924: [Feature Request] Adding an arc length constrain function
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Version

0.21 (Development)

### Full version info

```shell
[code]
OS: Manjaro Linux (KDE/plasma)
Word size of FreeCAD: 64-bit
Version: 0.21.0.33458 (Git) AppImage
Build type: Release
Branch: master
Hash: 8b89dd29d2382c72d738c52e8974216130ffa9db
Python 3.10.12, Qt 5.15.8, Coin 4.0.0, Vtk 9.1.0, OCC 7.6.3
Locale: German/Germany (de_DE)
Installed mods: 
  * PieMenu 1.2.7
  * sheetmetal 0.2.63
  * toSketch 1.0.1
  * A2plus 0.4.60n
  * QuickMeasure 2022.10.28
  * Curves 0.6.12
  * CfdOF 1.23.1
  * TabBar
  * IconThemes
  * Glass
  * ProDarkThemePreferencePack 1.0.0
[/code]
```

### Subproject(s) affected?

Sketcher

### Problem description

Hello all,
after some casual part making, I have discovered that FreeCAD is currently lacking an arc length constrain function.
Arc length constrains are mostly used for gears and other complex geometries, hence most CAD programs have it implemented in some form.
If you want to create an arc length constrain in FreeCAD, you are forced to use the Spreadsheet. Which, in my opinion, adds unnecessary complexity and feels disruptive to the overall workflow.

I would love to hear other opinions on this topic.
Many thanks in advance!

### Anything else?

_No response_

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

### PR DESCRIPTION
**Pull request by FlachyJoe**
Extend constraint distance to arc length.
Fix #9924 

Current status:
- [x] App part
image

- [x] Gui part

### CODE REVIEW

**Comment by FlachyJoe** [comment]
@PaddleStroke could you take a look at the GUI side? 

- When moving the label is only updates at mouse release .
- I'm using the `SoDatumLabel::ANGLE` but the hole between the arrows is too small.

image

**Comment by PaddleStroke** [comment]
For the first point, are others constraints moving correctly? Werner made a PR few days ago that could cause this exactly. He worked on undo/redo of movement of constraints which implied moveConstraint on mouseMove and release.

Yes the hole is a bit small, even for angles it is a bit small don't you think?

**Comment by PaddleStroke** [comment]
I'm not sure the symbol "◠" is necessary because the unit already makes it clear that it's a length no?

**Comment by FlachyJoe** [comment]
@PaddleStroke
The others constraints move as expected.
Shouldn't the label be above the line ?
The arc symbol requirement is explained here: https://github.com/FreeCAD/FreeCAD/issues/12036#issuecomment-1962661855

**Comment by PaddleStroke** [comment]
Ok so if it moves correctly, then the problem is probably in the moveConstraint function in VPSketch?

For the symbol, even when units are hidden, the deg is not hidden so I think it is still ok without symbol.
The thing is that I don't think the meaning of this arc symbol is so obvious at first glance (⭥ would be more clear imo).

Besides it clutters the view, especially for user who do not hide units.

Another solution would be to not hide the units of arc length even if the option is ON.

**Comment by PaddleStroke** [comment]
Regarding label above the line, why not, but then text has to be tangent to the arc. Which is not very readable in some cases

**Comment by NXTler** [comment]
I think we should follow ISO on the representation of arc lengths. It's simple and easy to differentiate: 
grafik
grafik

**Comment by FlachyJoe** [comment]
@NXTler :+1: the extends lines shouldn't be radial but parallel.
So the SoDatumLabel::ANGLE is not usable.

**Comment by PaddleStroke** [comment]
Ah ok if this symbol is a standard than it's just my ignorance

**Comment by maxwxyz** [comment]
just out of curiosity: this is for ArcOfCircle, not the other wild arc types? Or will it work with any non-linear line length?

**Comment by FlachyJoe** [comment]
image
Looks better but spacing has to be enlarged and I have an issue with the constraint selection: the active zone is not at the text location.

**Comment by FlachyJoe** [comment]
Quite ready but I have a problem with reference constraint which is taken into account by the solver:
image
image

**Comment by pierreporte** [comment]
@FlachyJoe Since you’re following ISO style for the arc length, would it be worth it to use the style for all other dimensions, for consistency? It would be out of scope of this PR though.

image

**Comment by FlachyJoe** [comment]
Ready for review and tests.

**Comment by oursland** [comment]
@FlachyJoe Moving the `_USE_MATH_DEFINES` macro to the top will fix the CI builds as shown [here](_USE_MATH_DEFINES).

The Windows Conda build is currently building with Visual Studio 2019 until the `conda-forge` team upgrades their `compilers` packages to a more recent edition, which is the reason the warnings and errors differ from the Windows (libpack) builds.

**Comment by oursland** [comment]
@chennes @adrianinsaval The issue previously preventing successful builds was resolved, however the Windows Conda CI is failing for some unknown reasons.  Can you retrigger that job?

**Comment by sliptonic** [comment]
@FlachyJoe  There seems to be a github issue.  It won't update the mergability.  Please force push against this branch to bump github

**Review by oursland — APPROVED** [review]

**Comment by FlachyJoe** [comment]
Thank you all for the support.

@PaddleStroke could you please add this arc length constraint to the Dimension tool?

**Comment by NXTler** [comment]
@pierreporte I opened a feature request for ISO standard constraints style #13392

**Comment by PaddleStroke** [comment]
@FlachyJoe I'm on it

**Comment by Roy-043** [comment]
Small issue: Currently ArcLength requires pre-selecting the object. If you start the Sketcher_ConstrainDistance tool first and then select there is a confusing warning.

```
OS: Windows 8 build 9600
Word size of FreeCAD: 64-bit
Version: 0.22.0dev.36807 (Git)
Build type: Release
Branch: main
Hash: cc96f2718e139d670e08c4226bdea9955a102ed6
Python 3.11.8, Qt 5.15.13, Coin 4.0.2, Vtk 9.2.6, OCC 7.7.2
Locale: Dutch/Netherlands (nl_NL)
Installed mods:
```

File: src/Gui/SoDatumLabel.cpp [hunk 1]
```diff
@@ -451,6 +522,7 @@ SbVec3f SoDatumLabel::getLabelTextCenter()
     const SbVec3f* points = this->pnts.getValues(0);
     SbVec3f p1 = points[0];
     SbVec3f p2 = points[1];
+    SbVec3f p3 = points[2];
```
Comment (wwmayer) [inline src/Gui/SoDatumLabel.cpp:525]: When building FreeCAD with the address sanitizer option I get a reliable crash here. Apparently the points field doesn't have four points.

Steps to reproduce:
* Create a sketch
* Try to create a rectangle
* It crashes when clicking inside the 3D view

Comment (FlachyJoe) [reply src/Gui/SoDatumLabel.cpp:525]: `points` has 3 fields only in `ARCLENGTH` case, it has 1 field in `ANGLE` one and 2 in others. A check can be added.

Comment (wwmayer) [reply src/Gui/SoDatumLabel.cpp:525]: https://github.com/FreeCAD/FreeCAD/pull/13627

**Pull request merged**